src/handlers/protect.js

```diff
@@ -0,0 +1,120 @@
+        const roles = token?.roles;
+        if (!roles) return redirect(config.redirect);
+        const roleNames = new Set(roles.map((r) => r.name));
+        if (!config.roles.some((role) => roleNames.has(role))) {
+          return redirect(config.redirect);
+        }
+      }
+
+      if (typeof config.permissions === 'string') {
+        const hasPermission = await getPermission(config.permissions);
+        if (!hasPermission) {
+          return redirect(config.redirect);
+        }
+      }
+
+      if (Array.isArray(config.permissions)) {
+        const permissions = await getPermissions();
+        if (
+          !config.permissions.some((permission) =>
+            permissions.includes(permission)
+          )
+        ) {
+          return redirect(config.redirect);
+        }
+      }
+    } catch (error) {
+      console.error('Error protecting page', error);
+      // return redirect(config.redirect);
+      return null;
+    }
+
+    return page(props);
+  };
+
+/**
+ * Protects a Next.js API route handler with authentication and authorization.
+ * @param {Function} handler - The Next.js API route handler.
+ * @param {Object} config - The configuration object.
+ * @param {string[]} config.roles - The required role(s) for accessing the protected page.
+ * @param {string|string[]} config.permissions - The required permission(s) for accessing the protected page.
+ * @returns {Function} - The protected API route handler.
+ */
+
+export const protectApi = (handler, config) => async (req) => {
+  const {isAuthenticated, getAccessToken, getPermission, getPermissions} =
+    kinde();
+  try {
+    const isSignedIn = await isAuthenticated();
+
+    if (!isSignedIn) {
+      return NextResponse.json({statusCode: 401, message: 'Unauthorized'});
+    }
+
+    if (config.roles) {
+      const token = await getAccessToken();
+      const roles = token?.roles;
+      if (!roles)
+        return NextResponse.json({statusCode: 401, message: 'Unauthorized'});
+      const roleNames = new Set(roles.map((r) => r.name));
+      if (!config.roles.some((role) => roleNames.has(role))) {
+        return NextResponse.json({statusCode: 401, message: 'Unauthorized'});
+      }
+    }
+
+    if (typeof config.permissions === 'string') {
+      const hasPermission = await getPermission(config.permissions);
+      if (!hasPermission) {
+        return NextResponse.json({statusCode: 403, message: 'Forbidden'});
+      }
+    }
+
+    if (Array.isArray(config.permissions)) {
+      const permissions = await getPermissions();
+      if (
+        !config.permissions.some((permission) =>
+          permissions.includes(permission)
+        )
+      ) {
+        return NextResponse.json({statusCode: 403, message: 'Forbidden'});
+      }
+    }
+  } catch (error) {
+    console.error('Error protecting page', error);
+    // return NextResponse.json({
+    //   statusCode: 500,
+    //   message: 'Internal Server Error'
+    // });
```

Comment:
Can we remove this comment please?

Reply:
The reason they're commented out is I wasn't sure how you would want to deal with errors from the functions. In other functions I saw you just returned null but maybe you would want to throw a 500.